[SPARK-18541][Python]Add metadata parameter to pyspark.sql.Column.alias()

shea-parkes opened the pull request:
## What changes were proposed in this pull request?

Add a `metadata` keyword parameter to `pyspark.sql.Column.alias()` to allow users to mix-in metadata while manipulating `DataFrame`s in `pyspark`.  Without this, I believe it was necessary to pass back through `SparkSession.createDataFrame` each time a user wanted to manipulate `StructField.metadata` in `pyspark`.  

This pull request also improves consistency between the Scala and Python APIs (i.e. I did not add any functionality that was not already in the Scala API).

Discussed ahead of time on JIRA with @marmbrus 

## How was this patch tested?

Added unit tests (and doc tests).  Ran the pertinent tests manually.

shea-parkes committed:
Add metadata arg to pyspark.sql.Column

Do this so it is possible to mix-in metadata during DataFrame manipulations.
Also do this to increase API parity with Scala.

AmplabJenkins commented:
Can one of the admins verify this patch?

marmbrus commented:
ok to test

SparkQA commented:
**[Test build #69517 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69517/consoleFull)** for PR 16094 at commit [`7af79a4`](https://github.com/apache/spark/commit/7af79a41c65c624158fbd9f15135f2e8d20ca936).

SparkQA commented:
**[Test build #69517 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69517/consoleFull)** for PR 16094 at commit [`7af79a4`](https://github.com/apache/spark/commit/7af79a41c65c624158fbd9f15135f2e8d20ca936).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69517/
Test FAILed.

shea-parkes commented:
Gah, I'm sorry.  I forgot python 2.6 compatibility.  I'll switch this over shortly to use the old-school style.  Likely do something like this:

```python
def alias(self, *alias, **kwargs):
    metadata = kwargs.pop('metadata', None)
    assert not kwargs
```

I'll also add some unit tests and raise a more proper exception (if I can find one).

Sorry again.

marmbrus commented:
No worries, thanks for working on this!  It's great to ensure our Python APIs aren't lagging behind the Scala ones.

shea-parkes committed:
Make keyword argument compatible with python 2.6

Do this because the straightforward way of doing a keyword only argument was not
introduced until Python 3.

SparkQA commented:
**[Test build #69527 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69527/consoleFull)** for PR 16094 at commit [`f9b8d5a`](https://github.com/apache/spark/commit/f9b8d5ae31fa8e45a2299f8b34933db3eae12a44).

shea-parkes commented:
Alright, I've pushed up a new commit that should make this compatible with python 2.6.  I don't have a convenient python 2.6 environment to test in, but it still is valid in my python 3 one.  I'll check back in on the CI results in a bit.

SparkQA commented:
**[Test build #69527 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69527/consoleFull)** for PR 16094 at commit [`f9b8d5a`](https://github.com/apache/spark/commit/f9b8d5ae31fa8e45a2299f8b34933db3eae12a44).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69527/
Test FAILed.

shea-parkes committed:
Correct assertRaises syntax.

Switch to the more proper context manager while I'm at it.

SparkQA commented:
**[Test build #69530 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69530/consoleFull)** for PR 16094 at commit [`cfa6274`](https://github.com/apache/spark/commit/cfa6274ad0e9402c52c59ee0bb55eaea44ff8447).

shea-parkes commented:
That last failure was a syntax screw-up in my intentionally failed unit tests; I'm too used to `py.test` style tests.  I've hopefully got the syntax correct now (and went ahead and switch to the cleaner context manager style while I was at it).  I'll check in again after the next round of CI results.

SparkQA commented:
**[Test build #69530 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69530/consoleFull)** for PR 16094 at commit [`cfa6274`](https://github.com/apache/spark/commit/cfa6274ad0e9402c52c59ee0bb55eaea44ff8447).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69530/
Test PASSed.

shea-parkes commented:
Alright @marmbrus , the tests are at least passing now (including the pesky python2.6).  Let me know if you'd like any further changes.

holdenk commented on python/pyspark/sql/column.py:
```diff
@@ -298,19 +299,34 @@ def isin(self, *cols):
         Returns this column aliased with a new name or names (in the case of expressions that
         return more than one column, such as explode).
 
+        Optional ``metadata`` keyword argument can be passed when aliasing a single column.
```
This is very minor, but it might make sense to have a versionChanged directive in the docString so its clear when metaData support was added (since its addiative and shouldn't break existing code more on the optional but possibly useful side).

shea-parkes replied:
I actually thought about this, but I'm struggling to find an example of how to properly document this.  I'll try grepping a little wider to see if I can find a "versionChanged" example somewhere...

shea-parkes replied:
Oh, and I inferred that this would be a v2.2 feature since v2.1 is already in pre-release.  If I got the version wrong, just let me know.

holdenk replied:
2.2 is probably right, although the current 2.1 RC is more a of a strawman so it is possible (but up to @davies / @marmbrus if this warrants going into 2.1).

shea-parkes replied:
Congrats on the release of v2.1.  I believe the `versionchanged` target of v2.2 is now accurate.

holdenk replied:
Excellent, so I think maybe using the `:param:` docstring notation here would be make sense (as in the other functions).

shea-parkes committed:
Added versionChanged directive to pyspark.sql.Column.alias

Do this so users are informed when this functionality became available.

shea-parkes commented:
There we go.  I found other examples of the `versionchanged` directive syntax and added a corresponding entry to `pyspark.sql.Column.alias`.  I think this should be good to look over again.  Let me know if there is anything else you would like added.

SparkQA commented:
**[Test build #69602 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69602/consoleFull)** for PR 16094 at commit [`3711ca7`](https://github.com/apache/spark/commit/3711ca72a14e019200358f1d5fc5149010e4ceb2).

SparkQA commented:
**[Test build #69602 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69602/consoleFull)** for PR 16094 at commit [`3711ca7`](https://github.com/apache/spark/commit/3711ca72a14e019200358f1d5fc5149010e4ceb2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69602/
Test PASSed.

holdenk commented:
Lets see if maybe @marmbrus or @davies has some time to look at this?

shea-parkes commented:
I'm sure everyone was busy over the holidays and with the release of v2.1 (congrats).

Anecdotally, we've been using this functionality a *lot* in our internal fork, so I think it could be quite valuable to get more broadly available.  It also exposes more of the Scala API to Python.

I'd be happy to make any requested changes.  Just let me know what you'd like done.

shea-parkes commented:
Looks like v2.2 hasn't been branched off yet.  We're still using the dickens out of this feature locally.  It's also a good step forward for Python API parity.  Is there anyone with some time to look at this?

I'd be happy to make any further requested changes, although I think it's in a good state now.  Thanks.

holdenk commented:
I'm pretty busy in the run up to Spark Summit - let's chat at the summit if you are around or I can take a look at this after once my schedule is back to normal. Sorry for the delay in getting to this @shea-parkes & thanks for sticking with it.

holdenk reviewed (commented):
This look reasonable, one minor point with maybe improving the docstring a bit (which will also give us a fresh jenkins test since its been awhile). If you have chance to update the docstring and jenkins passes I'll try and merge this :)

shea-parkes committed:
Enhance Column.alias docstring.

Do this to better conform to param documentation standards.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #72871 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72871/testReport)** for PR 16094 at commit [`6305ab4`](https://github.com/apache/spark/commit/6305ab425609f4377ec2733655a0537e1e5c6557).

SparkQA commented:
**[Test build #72871 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72871/testReport)** for PR 16094 at commit [`6305ab4`](https://github.com/apache/spark/commit/6305ab425609f4377ec2733655a0537e1e5c6557).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72871/
Test PASSed.

holdenk commented:
Thank you for updating this, looks good to me! :)

holdenk commented:
Merged to master :)
